src/VisualStudio/Core/Def/Implementation/Diagnostics/VisualStudioDiagnosticAnalyzerService.cs

```diff
@@ -199,29 +199,25 @@ public void RunAnalyzers(IVsHierarchy? hierarchy)
             string? projectOrSolutionName = project?.Name ?? PathUtilities.GetFileName(solution.FilePath);
 
             // Add a message to VS status bar that we are running code analysis.
-            var statusMessage = projectOrSolutionName != null
-                ? string.Format(ServicesVSResources.Running_code_analysis_for_0, projectOrSolutionName)
-                : ServicesVSResources.Running_code_analysis_for_Solution;
             var statusBar = _serviceProvider?.GetService(typeof(SVsStatusbar)) as IVsStatusbar;
-            statusBar?.SetText(statusMessage);
+            var totalProjectCount = project != null ? 1 : (uint)solution.ProjectIds.Count;
+            var statusBarUpdater = statusBar != null ?
+                new StatusBarUpdater(statusBar, _threadingContext, projectOrSolutionName, totalProjectCount) :
+                null;
 
             // Force complete analyzer execution in background.
             var asyncToken = _listener.BeginAsyncOperation($"{nameof(VisualStudioDiagnosticAnalyzerService)}_{nameof(RunAnalyzers)}");
             Task.Run(async () =>
             {
-                await _diagnosticService.ForceAnalyzeAsync(solution, project?.Id, CancellationToken.None).ConfigureAwait(false);
+                var onProjectAnalyzed = statusBarUpdater != null ? statusBarUpdater.OnProjectAnalyzed : (Action<Project>)((Project _) => { });
+                await _diagnosticService.ForceAnalyzeAsync(solution, onProjectAnalyzed, project?.Id, CancellationToken.None).ConfigureAwait(false);
 
                 // If user has disabled live analyzer execution for any project(s), i.e. set RunAnalyzersDuringLiveAnalysis = false,
                 // then ForceAnalyzeAsync will not cause analyzers to execute.
                 // We explicitly fetch diagnostics for such projects and report these as "Host" diagnostics.
                 HandleProjectsWithDisabledAnalysis();
 
-                // Add a message to VS status bar that we completed executing code analysis.
-                await _threadingContext.JoinableTaskFactory.SwitchToMainThreadAsync();
-                var notificationMesage = projectOrSolutionName != null
-                    ? string.Format(ServicesVSResources.Code_analysis_completed_for_0, projectOrSolutionName)
-                    : ServicesVSResources.Code_analysis_completed_for_Solution;
-                statusBar?.SetText(notificationMesage);
+                statusBarUpdater?.Dispose();
```

Comment:
what happens if we cancel/throw?  will this not be disposed?  what's the impact of tht?

Reply:
I can wrap this in a try/finally and move the dispose call into finally.